src/GafferImage/ImageStats.cpp

```diff
@@ -198,98 +222,219 @@ const DeepState *ImageStats::deepState() const
+			Imath::Box2i( boundsIntersection.min - tileOrigin, boundsIntersection.max - tileOrigin ),
+			Imath::Box2i( Imath::V2i( 0 ), Imath::V2i( ImagePlug::tileSize() ) )
+		);
+		h.append( tileBound );
+		flattenedInPlug()->channelDataPlug()->hash( h );
 	}
+	else if( output == allStatsPlug() )
+	{
+		if( BufferAlgo::empty( boundsIntersection ) )
+		{
+			h.append( 0.0f );
+			return;
+		}
 
-	Sampler s( inPlug(), channelName, area );
-	s.hash( h );
+		ImageAlgo::parallelGatherTiles(
+			flattenedInPlug(),
+			// Tile
+			[this] ( const ImagePlug *imageP, const Imath::V2i &tileOrigin )
+			{
+				return tileStatsPlug()->hash();
+			},
+			// Gather
+			[ &h ] ( const ImagePlug *imageP, const Imath::V2i &tileOrigin, const IECore::MurmurHash &tileHash )
+			{
+				h.append( tileHash );
+			},
+			boundsIntersection,
+			ImageAlgo::BottomToTop
+		);
+		h.append( areaMult );
+	}
 }
 
 void ImageStats::compute( ValuePlug *output, const Context *context ) const
 {
-	const int colorIndex = ::colorIndex( output );
-	if( colorIndex == -1 )
+	const Plug *parent = output->parent<Plug>();
+	if(
+		parent == minPlug() ||
+		parent == maxPlug() ||
+		parent == averagePlug()
+	)
 	{
-		// Not a plug we know about
-		ComputeNode::compute( output, context );
+		const int colorIndex = ::colorIndex( output );
+		if( colorIndex == -1 )
+		{
+			// Not a plug we know about
+			return;
+		}
+		const std::string channelName = this->channelName( colorIndex );
+		if( channelName.empty() )
+		{
+			static_cast<FloatPlug *>( output )->setValue( 0.0f );
+			return;
+		}
+
+		int statIndex = ( parent == averagePlug() ) ? 2 : ( parent == maxPlug() );
+
+		ImagePlug::ChannelDataScope s( context );
+		s.setChannelName( channelName );
+		Imath::V3f stats = allStatsPlug()->getValue().min;
+		static_cast<FloatPlug *>( output )->setValue( stats[ statIndex ] );
 		return;
 	}
 
-	const std::string channelName = this->channelName( colorIndex );
-	const Imath::Box2i area = areaPlug()->getValue();
+	Imath::Box2i boundsIntersection;
+	double areaMult;
 
-	if( channelName.empty() || BufferAlgo::empty( area ) )
 	{
-		static_cast<FloatPlug *>( output )->setValue( 0.0f );
-		return;
+		ImagePlug::GlobalScope s( context );
+		const Imath::Box2i area = areaPlug()->getValue();
+		const Imath::Box2i dataWindow = flattenedInPlug()->dataWindowPlug()->getValue();
+		boundsIntersection = BufferAlgo::intersection( area, dataWindow );
+		areaMult = double(area.size().x) * area.size().y;
 	}
 
-	// Loop over the ROI and compute the min, max and average channel values and then set our outputs.
-	Sampler s( inPlug(), channelName, area );
+	if( output == tileStatsPlug() )
+	{
+		Imath::V2i tileOrigin = context->get<Imath::V2i>( ImagePlug::tileOriginContextName );
+		const Imath::Box2i tileBound = BufferAlgo::intersection(
+			Imath::Box2i( boundsIntersection.min - tileOrigin, boundsIntersection.max - tileOrigin ),
+			Imath::Box2i( Imath::V2i( 0 ), Imath::V2i( ImagePlug::tileSize() ) )
+		);
```

Comment:
> Implement the final computation using parallelGatherTiles() or parallelProcessTiles(), pulling on the internal tile-stats plug for any tiles completely overlapped by the sampling region. (from original issue description)

I don't think I communicated this very well. I was thinking that `tileStatsPlug()` would contain the stats for the entire valid region of the tile, not the region clipped by `areaPlug()`. This would make interactive updates much quicker when changing the area, because we could reuse the stats for all tiles which are completely enclosed by the area. Only the final `allStatsPlug()` compute would depend on area, and that would be responsible for dealing with the clipped tiles at the edges.

Reply:
I'd misread this - we're not including `areaPlug()` directly in the hash, but rather it's intersection with the tile bound. So we are indeed able to reuse cached results, as demonstrated in [this new test](https://github.com/GafferHQ/gaffer/pull/3792/commits/b490ad0c911f47d7e913154008db09cea6371bff#diff-441a2914869fe7e350ab2ba455418e1fR166).